callback_plugins/custom_junit.py

```diff
@@ -0,0 +1,34 @@
+from ansible.plugins.callback.junit import CallbackModule as JunitCallbackModule
+
+import os
+import re
+
+class CallbackModule(JunitCallbackModule):
+    CALLBACK_NAME = 'custom_junit'
+
+    def __init__(self):
+        super(CallbackModule, self).__init__()
+
+        # Enforce our own defaults for when ENVVARs are inconvenient
+        self._output_dir = os.getenv('JUNIT_OUTPUT_DIR', '/var/lib/AnsibleTests/external_files/')
+        self._test_case_prefix = os.getenv('JUNIT_TEST_CASE_PREFIX', 'rhelai-validation : TEST')
+        self._fail_on_ignore = os.getenv('JUNIT_FAIL_ON_IGNORE', 'true')  # this is needed because we use "ignore_errors" on assertion tasks to run as many as possible
+        self._hide_task_arguments = os.getenv('JUNIT_HIDE_TASK_ARGUMENTS', 'true')
+
+    def mutate_task_name(self, task_name):
+        if len(self._test_case_prefix) > 0:
+            task_name = task_name.split(self._test_case_prefix)[-1]  # remove the test prefix and everything before it
+        task_name = task_name.lower()
+        task_name = re.sub(r'\W', ' ', task_name)  # replace all non-alphanumeric characters (except _) with a space
+        task_name = re.sub(r'(^\W*|\W*$)', '', task_name)  # trim any trailing or leading non-alphanumeric characters
+        task_name = re.sub(r' +', '_', task_name)  # replace any number of spaces with _
+        return task_name
+
+    def _build_test_case(self, task_data, host_data):
+        tc = super()._build_test_case(task_data, host_data)
+
+        tc.name = self.mutate_task_name(tc.name)
+        tc.system_out = None
+        tc.system_err = None
+        tc.classname = 'rhoso_rhelai_validation.' + re.sub(r'\.yaml:[0-9]+$', '', os.path.basename(task_data.path))
```

Comment:
I don't really like this classname system (that I just finished writing). It means that the classname will change if we reorganize our ansible code, which doesn't seem ideal to me (though I know historically that's what classname means). Instead I was thinking of something like "TEST[gpus]"  "TEST[nvidia]", etc and I could parse that class name out to add to `rhoso_rhelai_validation.`

Thoughts?

Reply:
Output snippet for context:
```
<testcase classname="rhoso_rhelai_validation.nvidia_tests" name="check_if_gpus_in_passthrough_mode_are_present_in_rhel_ai_vm_nvidia_smi" time="0.025861740112304688"/>
```

I would like it if the classname + name uniquely identified the test, wherever it may move in a refactored codebase (like your UUID). For example, if we decide naming our taskfiles `*_asserts.yaml` is better than `*_tests.yaml` I don't want to have to adjust any downstream testing systems. Right now the classname suffix depends on the path, so while I don't think we need to "add the full name into the test task", explicitly adding the classname suffix seems like a good idea. I'll stop arguing against my own diff and instead code up my proposed improvement. :smile: